File: BAEKJOON/14891 - 톱니바퀴/톱니바퀴.cs
using System;
using System.Collections.Generic;

class Program
{
    // 톱니바퀴 회전 함수
    static List<int> RotateGear(List<int> gear, int direction)
    {
        // 시계 방향 회전
        if (direction == 1)
        {
            // 마지막 원소를 맨 앞으로 이동
            int last = gear[gear.Count - 1];
            gear.RemoveAt(gear.Count - 1);
            gear.Insert(0, last);
        }
        // 반시계 방향 회전
        else if (direction == -1)
        {
            // 첫 번째 원소를 맨 뒤로 이동
            int first = gear[0];
            gear.RemoveAt(0);
            gear.Add(first);
        }
        return gear;
    }

    // 톱니바퀴 회전 시뮬레이션 함수
    static void SimulateGears(List<List<int>> gears, List<Tuple<int, int>> rotations)
    {
        foreach (var rotation in rotations)
        {
            int gearIndex = rotation.Item1 - 1;
            int direction = rotation.Item2;

            // 각 톱니바퀴의 회전 방향을 기록하는 배열
            int[] rotationDirections = new int[4];
            rotationDirections[gearIndex] = direction;

            // 현재 톱니바퀴의 왼쪽으로 영향 전파
            for (int i = gearIndex; i > 0; i--)
            {
                if (gears[i][6] != gears[i - 1][2])
                {
                    rotationDirections[i - 1] = -rotationDirections[i];
                }
                else
                {
                    break;
                }
            }

            // 현재 톱니바퀴의 오른쪽으로 영향 전파
            for (int i = gearIndex; i < 3; i++)
            {
                if (gears[i][2] != gears[i + 1][6])
                {
                    rotationDirections[i + 1] = -rotationDirections[i];
                }
                else
                {
                    break;
                }
            }

            // 기록된 회전 방향에 따라 톱니바퀴 회전
            for (int i = 0; i < 4; i++)
            {
                if (rotationDirections[i] != 0)
                {
                    gears[i] = RotateGear(gears[i], rotationDirections[i]);
                }
            }
        }
    }

    // 점수 계산 함수
    static int CalculateScore(List<List<int>> gears)
    {
        int score = 0;
        for (int i = 0; i < 4; i++)
        {
            // 12시 방향이 S극인 경우
            if (gears[i][0] == 1)
            {
                score += (1 << i); // 2의 i승 더하기
            }
        }
        return score;
    }

    static void Main()
    {
        List<List<int>> gears = new List<List<int>>();
        List<Tuple<int, int>> rotations = new List<Tuple<int, int>>();

        // 톱니바퀴 초기 상태 입력
        for (int i = 0; i < 4; i++)
        {
            string input = Console.ReadLine();
            List<int> gear = new List<int>();
            foreach (char c in input)
            {
                gear.Add(c - '0'); // 문자열을 숫자로 변환하여 저장
            }
            gears.Add(gear);
        }

        int K = int.Parse(Console.ReadLine()); // 회전 횟수 입력

        // 회전 명령 입력
        for (int i = 0; i < K; i++)
        {
            string[] rotationInfo = Console.ReadLine().Split();
            int gearIndex = int.Parse(rotationInfo[0]);
            int direction = int.Parse(rotationInfo[1]);
            rotations.Add(Tuple.Create(gearIndex, direction));
        }

        // 시뮬레이션 실행
        SimulateGears(gears, rotations);

        // 점수 계산 및 출력
        int score = CalculateScore(gears);
        Console.WriteLine(score);
    }
}

File: BAEKJOON/14891 - 톱니바퀴/톱니바퀴.py
def rotate_gear(gear, direction):
    # 톱니바퀴를 시계 방향으로 회전
    if direction == 1:
        # 톱니바퀴를 오른쪽으로 한 칸씩 이동
        return [gear[-1]] + gear[:-1]
    # 톱니바퀴를 반시계 방향으로 회전
    elif direction == -1:
        # 톱니바퀴를 왼쪽으로 한 칸씩 이동
        return gear[1:] + [gear[0]]

def simulate_gears(gears, rotations):
    # 주어진 회전 명령들을 순차적으로 처리
    for gear_index, direction in rotations:
        # 각 톱니바퀴의 회전 방향을 저장하는 리스트 초기화
        rotation_directions = [0, 0, 0, 0]
        # 현재 회전시킬 톱니바퀴의 방향 설정
        rotation_directions[gear_index - 1] = direction
        
        # 현재 톱니바퀴의 왼쪽에 있는 톱니바퀴들에 대해 회전 영향 전파
        for i in range(gear_index - 1, 0, -1):
            # 현재 톱니바퀴와 왼쪽 톱니바퀴의 맞닿은 극이 다를 경우
            if gears[i][6] != gears[i - 1][2]:
                # 왼쪽 톱니바퀴는 현재 톱니바퀴와 반대 방향으로 회전
                rotation_directions[i - 1] = -rotation_directions[i]
            else:
                # 맞닿은 극이 같으면 더 이상 왼쪽으로 전파되지 않음
                break
        
        # 현재 톱니바퀴의 오른쪽에 있는 톱니바퀴들에 대해 회전 영향 전파
        for i in range(gear_index - 1, 3):
            # 현재 톱니바퀴와 오른쪽 톱니바퀴의 맞닿은 극이 다를 경우
            if gears[i][2] != gears[i + 1][6]:
                # 오른쪽 톱니바퀴는 현재 톱니바퀴와 반대 방향으로 회전
                rotation_directions[i + 1] = -rotation_directions[i]
            else:
                # 맞닿은 극이 같으면 더 이상 오른쪽으로 전파되지 않음
                break
        
        # 저장된 회전 방향에 따라 톱니바퀴들을 회전
        for i in range(4):
            if rotation_directions[i] != 0:
                gears[i] = rotate_gear(gears[i], rotation_directions[i])
    
    # 회전 후의 톱니바퀴 상태 반환
    return gears

def calculate_score(gears):
    # 최종 점수를 저장할 변수 초기화
    score = 0
    # 각 톱니바퀴의 12시 방향을 검사하여 점수 계산
    for i in range(4):
        if gears[i][0] == '1':
            # 톱니바퀴의 12시 방향이 S극('1')이면 점수를 추가
            score += 2 ** i
    # 최종 점수 반환
    return score

def main():
    import sys
    input = sys.stdin.read
    data = input().split()
    
    # 첫 4개의 줄을 통해 톱니바퀴의 초기 상태를 저장
    gears = [list(data[i]) for i in range(4)]
    
    # 회전 횟수 K를 입력받음
    K = int(data[4])
    
    # 회전 정보들을 저장할 리스트 초기화
    rotations = []
    # 회전 정보는 톱니바퀴 번호와 회전 방향으로 주어지므로 이를 파싱하여 저장
    for i in range(5, 5 + K * 2, 2):
        gear_index = int(data[i])
        direction = int(data[i + 1])
        rotations.append((gear_index, direction))
    
    # 주어진 회전 정보에 따라 톱니바퀴들을 회전시킴
    gears = simulate_gears(gears, rotations)
    
    # 회전이 모두 끝난 후 최종 점수를 계산
    score = calculate_score(gears)
    
    # 최종 점수 출력
    print(score)

if __name__ == "__main__":
    main()
